### ISSUE TITLE
Solution for enrypted Message

### ISSUE THREAD

**Issue by ilhamiarie** [gi097/WhatsApi-Android#5]
Hai @gi097, is there a solution to decrypt the encrypted message ? i've run your latest code, but stil can not. thanks.

**Comment by gi097** [comment]
Nope, I can't fix it since I don't know how to do it. Sometimes it is working, but there is something wrong in adjustId() and deAdjustId(), @mgp25 is taking a look at it when he got time.

**Comment by ilhamiarie** [comment]
hopefully @mgp25 can spare a little time to see it. Thanks a lot.

**Comment by mgp25** [comment]
Im not a very experienced java user, anyway take this as reference:

```java
public function adjustId(byte[] id)
{
  byte[] data = new StringBuilder(new String(Hex.encodeHex(id))).reverse().toString();
  data = bin2hex(data)
  // ltrim its done like this:
  // String ltrim = s.replaceAll("^\\s+","");
...
...
}


  public static String bin2Hex(byte bytes[]) 
  {
    StringBuffer retString = new StringBuffer();
    for (int i = 0; i < bytes.length; ++i)
    {
      retString.append(
      Integer.toHexString(0x0100 + (bytes[i] & 0x00FF)).substring(1));
    }
    return retString.toString();
  }
```

Hope this can help you to continue your project

Regards

**Comment by gi097** [comment]
@mgp25, thank you! I think I can do something with this. But, where is deAdjustId()?

**Comment by ilhamiarie** [comment]
function deAdjustId($data){
  if(strlen($data) < 4)
    $data = "\x00".$data;
    $data = strrev($data);
    $data = unpack("L",$data);
    return $data[1];
}
the problem ( pack and unpack function) in java format. i couldn find.this solution. 
cc. @mgp25 , @gi097 

Regards.

**Comment by gi097** [comment]
Yeah same here, I got everything except the pack and unpack functions.

**Comment by gi097** [comment]
@ilhamiarie you can try this:

    public static byte[] adjustId(int id) {
        String data = BinHex.bin2hex(new StringBuilder(new String(Hex.encodeHex(ByteBuffer.allocate(4).putInt(id).array()))).reverse().toString().replaceAll("^\\s+", "").getBytes());
        while (data.length() < 6) {
            data = "0" + data;
        }
        return BinHex.hex2bin(data);
    }

    public static int deAdjustId(byte[] id) {
        String data = new String(id);
        if (data.length() < 4) {
            data = "\\x00" + data;
        }
        data = new StringBuilder(data).reverse().toString();
        try {
            byte[] unpack = Hex.decodeHex(data.toCharArray());
            return (unpack[1] & 0xFF) << 16;
        } catch (DecoderException e) {
            e.printStackTrace();
            return 0;
        }
    }
